NOTES-CMAKE.md

```diff
@@ -0,0 +1,90 @@
+Notes for CMake generator
+===========================
+
+Requirement details
+-------------------
+
+In addition to the obvious CMake requirement, these are required as well:
+
+## Perl
+
+We recommend Strawberry Perl, or alternatively ActiveState Perl on Windows.
+Use the appropriate package manager for Unix systems to acquire Perl.
+
+please see the [Notes for Windows platforms](NOTES-WINDOWS.md) or [Notes for Unix platforms](NOTES-UNIX.md).
+
+
+## Windows
+
+### Netwide Assembler (NASM) on Windows
+
+NASM is the only currently supported assembler on Windows. It is available from <https://www.nasm.us> and has great support with CMake.
+
+Quick start
+-----------
+
+ 1. Install Perl (and NASM is necessary)
+
+ 3. Make sure CMake, Perl, and NASM are all on your PATH.
+
+ 5. From the root of the OpenSSL source directory enter
+    - `perl Configure CMake`
+
+    The perl Configure script will produce `openssl-config.cmake` in the root OpenSSL source directory.
+
```

Comment:
This is very weird. CMake is supposed to be "most external" wrapper and here we use Configure script to generate cmake-config... I think openssl-config.cmake should be present in the repo from the start?
This step should be handled by end-user with something like [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html) (specifically [FetchContent_MakeAvailable()](https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_makeavailable)). It always amazes me how [GTest](https://google.github.io/googletest/quickstart-cmake.html) does it...

Reply:
We (the OpenSSL team, and quite strongly, I) have pushed back on having CMake as the "most external" wrapper, because it means that we lose some platforms that are currently supported.  This has been reasoned multiple time already in diverse issues on the matter.

Reply:
Even with `Configure` as the "most important" tool, I think that CMake should still be the "most external" wrapper for the plaform which it supports, passing configuration into `Configure` and then loading the generated result.

Reply:
@dg0yt, I actually disagree with that view for the simple reason, the installation instruction becomes ambiguous and thereby more confusing.

"On some platforms, use cmake, on others, use Configure directly"

or

"Try cmake...  if you're unlucky and it doesn't work right, run Configure directly instead"

Reply:
@levitte I understand this point of view. However it remains quite abstract which platforms are actually affected.

The other approach creates ambiguity, too. First I need to pass toolchain and target related properties into `Configure`, and then I pass toolchain and target related properties into CMake. This is easy to get wrong.

That's the core of the problem: both `Configure` and CMake (or meson, or...) take toolchain and target related properties, and generate a build system. The plus side of CMake (or meson, or ...) is that it is widely known and well integrated into package managers and IDEs, and it  supports modern build tools (in particula ninja). The plus side of `Configure` is that it is exactly tailored to your needs and that it supports some platforms which are assumed to be reachable via CMake.

If you only want CMake config files, you don't need CMake.
If you only want ninja build files, you don't need CMake.
In the end, `Configure` solves the same problems which CMake and other tools solve, but in a proprietary way.

Reply:
Yes, exactly. Our stuff was made to do the same as CMake, more or less (as a matter of fact, I've been inspired by CMake on multiple occasions), but made to work on a wider range of platforms.
Whichever way you do things, it's a bit weird to do it so one is a subordinate of the other.

Side note: I object to our build system being proprietary. The source is all there, you can read it. And just so we compares apples with apples, I don't find the CMake source particularly easy to read, especially for figuring out how to make a new generator.

Reply:
> That's the core of the problem: both Configure and CMake (or meson, or...) take toolchain and target related properties, and generate a build system.

That is indeed the core of the problem -- this is a misunderstanding.

Both Configure and CMake or meson (or or or... ) are build systems, they don't generate build systems. They generate command runners (Makefile, build.ninja).

The goal of this PR is to have one build system (Configure) generate another buildsystem (CMake), for the preference of people that happen to like CMake. In this respect, it's probably closer to premake, which has modules to export cmake files (and also exports several other formats natively), and each premake export format is standalone and knows how to both configure and build itself.

If all the logic to define what cmake code should even be run in the first place, is going to reside in Configure, then it simply makes basic sense that one must call Configure before calling cmake, and thus, Configure *must* be the top-level controller.

Reply:
I notice that terminology can be understood more sensitively that it was meant.

By "proprietary", I was focusing on the fact how `Configure` is unique to this project.
I'm fine with using "command runner" when I used "build system".

> If all the logic to define what cmake code should even be run in the first place, is going to reside in Configure, then it simply makes basic sense that one must call Configure before calling cmake, and thus, Configure must be the top-level controller.

Maybe my point becomes clearer when I avoid the term "top-level". I don't want to propose to change the *central* configuration system. I want to ensure that `CMakeLists.txt` can act as an *entry point* to configuration as much as in any other project that is using CMake. This would be a benefit for packagers which can apply there existing routines as well as for users which prefer integration via CMake's  `add_subdirectory(openssl)`.

The approach in this PR has the potential to deliver this user experience, and it does so by integrating the execution of `Configure` into CMakeList.txt.

If CMake is degraded to an intermediate "command runner" where you cannot do typical CMake interactions because you are actually supposed to manually re-run Configure, I expect different types of issues on the user side.

Reply:
> The goal of this PR is to have one build system (Configure) generate another buildsystem (CMake), for the preference of people that happen to like CMake.

This looks like an opinion of a person that has great knowledge about build systems but not exactly on CMake in particular (in concrete - how CMake "packaging" works). By word "packaging" I mean the process of discovery CMake project A by the CMake project B.
I will quote some sentences from this [CMake Using dependencies guide](https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#guide:Using%20Dependencies%20Guide) to explain to you why it would be better to have CMake as "side by side" system to Configure file, not as "child of Configure" build system.

> "The primary methods of bringing dependencies into the build are the [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) command and the [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html#module:FetchContent) module."

> A package needed by the project may already be built and available at some location on the user's system. That package might have also been built by CMake, or it could have used a different build system entirely. It might even just be a collection of files that didn't need to be built at all. CMake provides the [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) command for these scenarios. It searches well-known locations, along with additional hints and paths provided by the project or user.

As you can see, CMake is flexible in finding packages. **Our goal is to make it a convenient process.**
Ok, after some introduction - the last sentence is the one I would like to elaborate on:
_It searches well-known locations, along with additional hints and paths provided by the project or user._
Now, from [find_package](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) documentation:

> The command has a few modes by which it searches for packages: Module mode and Config mode

I will not get onto details or differences between those modes but the working mechanism is similar -
a **CMake project has a specifically named file present in its directory** tree (eg. OpenSslConfig.cmake).
The presence of those files allows us to "Runtime configure some UserProjectA & OpenSSL project that is a dependency to it" (using [FetchContent_declare()](https://cmake.org/cmake/help/latest/module/FetchContent.html#command:fetchcontent_declare)). I mean following flow:
1) User has CMake projectA that uses OpenSSL project.
User executes"[cmake configure](https://cmake.org/cmake/help/latest/manual/cmake.1.html#generate-a-project-buildsystem)" step of HIS project. Let's say `cmake -S projectA -B build`.
2) In this very moment, OpenSSL project is cloned to some directory of projectA, is configured also using "[cmake configure](https://cmake.org/cmake/help/latest/manual/cmake.1.html#generate-a-project-buildsystem)" step. Moreover, configuration of the OpenSSL project matches end-user expectations. OpenSSL is configured for some desired architecture X to be built with [build generator](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html) Y.
3) User calls build command on his projectA (typically `cmake --build build`).
This builds projectA and OpenSSL. End-user may define dependencies ob build time in projectA - when he/she wants
to build OpenSSL (order), which targets of projectA actually need OpenSSL and many more.

If we do not provide a predefine CMake file for dependency discovery (OpenSslConfig.cmake) from the very start in the repo, process described above is not possible. So the solution with Configure script generating a Cmake module file has many disadvantages.

### Further consequences
Let's say that we ignore what I'm suggesting and proceed with Configure script idea.
Let's analyze the case where we build single OpenSSL project in our operating system and use it for many projects (projectA, projectB).
We clone OpenSSL to some directory and run Configure cmake to generate OpenSslConfig.cmake file and CMakeLists.txt files. This would probably also register OpenSSL project to so-called [cmake package registry](https://cmake.org/cmake/help/latest/manual/cmake-packages.7.html#package-registry) (which is a system registry on Windows, hidden ~/.cmake directory on linux etc.). Registration to the package registry allows having an OpenSSL path found by other projects. ProjectA expects to build OpenSSL for x64 while ProjectB cross-compiles OpenSSL for ARM with some custom toolchain. Building projectA built OpenSSL for x64. Now, end-user does not clean anything in OpenSSL directory and just build projectB that simply looks for libs or executable in the OpenSSL directory. I'm afraid that this may lead to using wrong architecture files/artifacts... Why? Because if you disable flow with FetchContent, you encourage users to use CMake this way (single project installation in the OS).

Reply:
> ... I will quote some sentences from this [CMake Using dependencies guide](https://cmake.org/cmake/help/latest/guide/using-dependencies/index.html#guide:Using%20Dependencies%20Guide) to explain to you why it would be better to have CMake as "side by side" system to Configure file, not as "child of Configure" build system.
> 
> > "The primary methods of bringing dependencies into the build are the [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) command and the [FetchContent](https://cmake.org/cmake/help/latest/module/FetchContent.html#module:FetchContent) module."
> 
> > A package needed by the project may already be built and available at some location on the user's system. That package might have also been built by CMake, or it could have used a different build system entirely. It might even just be a collection of files that didn't need to be built at all. CMake provides the [find_package()](https://cmake.org/cmake/help/latest/command/find_package.html#command:find_package) command for these scenarios. It searches well-known locations, along with additional hints and paths provided by the project or user.

This actually suggests that making OpenSSL buildable with cmake isn't necessary at all...  all that's necessary would be that the build generates an appropriate OpenSSLConfig.cmake and OpenSSLConfigVersion.cmake, and require that `find_package` be used in config mode by other projects.  *It is not necessary to use cmake to generate those files!*  We already do this sort of thing, but for `pkg-config` (see the files `$PREFIX/lib/pkgconfig/*.pc` that are installed with any OpenSSL installation on Unix family platforms); it would probably not be terribly hard to do it for CMake `find_package` support as well, both build-tree support (i.e. what the CMake `export()` command does) and install tree support (i.e. what `install(EXPORT ...)` does).

I have this sort of support in the works, just gotta make it good enough to submit a PR.

Reply:
For the record, I'm quite familiar with how cmake packages work. Disagreement about the correct way to do things doesn't imply lack of knowledge about how cmake currently implements it.

> ProjectA expects to build OpenSSL for x64 while ProjectB cross-compiles OpenSSL for ARM with some custom toolchain. Building projectA built OpenSSL for x64. Now, end-user does not clean anything in OpenSSL directory and just build projectB that simply looks for libs or executable in the OpenSSL directory. I'm afraid that this may lead to using wrong architecture files/artifacts... Why? Because if you disable flow with FetchContent, you encourage users to use CMake this way (single project installation in the OS).

This is incorrect, or rather, it's based on a misunderstanding of how environments work.


You want to encourage users to use single project installation in the OS.

You also want to *discourage* uses of package registries that don't distinguish between native and cross compiler toolchains.

Particularly, cmake has a native search path for package configs, and it is ***crucial*** that this be turned off when cross compiling, by replacing it with an alternative cross compile root. This is not automatically the case.

Likewise, you must turn off the package registry. Most people will forget to do this.

Do not under any circumstances add cross compiled project installations to the cmake package registry.

...

TBH, cmake is just not very good at cross compiling except in the simplest cases (see for example the fact that it cannot handle try_run, or the way building codegen tools requires manually running cmake twice, with ifelse constructs to either build a native codegen tool or else find the manually built one, depending on whether the cmake is cross-compiling at the moment.

I would use a better build system if you need to cross compile, but assuming that isn't an option... cross compiling is complicated at the best of times and it's the responsibility of a person doing so to use best practices for environment setup so that package detection isn't broken.

(Meson has first class cross compilation support. You can tag an executable as native, which means it's always built with a native compiler even during a cross build, and that executable then runs automatically as part of the same basic build setup. Package detection does NOT work in a cross compile until you tell it how to do so safely. This includes both setting up pkg-config and setting up cmake find_package, which yes, meson supports.)